dhall/src/Dhall/Import.hs

```diff
@@ -463,88 +465,376 @@ localToPath prefix file_ = liftIO $ do
-    let Import {..} = here
-
-    let ImportHashed {..} = importHashed
-
+            semanticImport <- processImportWithSemanticCache stack import_
+            -- Lens.Family.State.Strict only exports `%=`, not `modifying`...
+            cache %= (Map.insert import_ semanticImport)
+            return semanticImport
+
+-- Checks the "semantic" disk cache; defers to
+-- `processImportWithSemisemanticCache` in general.
+processImportWithSemanticCache
+  :: ImportStack -> ResolvedImport -> StateT Status IO SemanticImport
+-- Pass on non-frozen imports.
+processImportWithSemanticCache stack import_@(ResolvedImport Nothing _ _) =
+    processImportWithSemisemanticCache stack import_
+
+processImportWithSemanticCache stack
+  import_@(ResolvedImport (Just semanticHash) _ _) = do
+    mCached <- lift $ fetchFromSemanticCache semanticHash
+    case mCached of
+        Just bytesStrict -> do
+            let actualHash = Crypto.Hash.hash bytesStrict
+            if semanticHash == actualHash
+                then return ()
+                else throwMissingImport (Imported stack (HashMismatch {expectedHash = semanticHash, ..}))
+
+            let bytesLazy = Data.ByteString.Lazy.fromStrict bytesStrict
+            term <- Dhall.Core.throws (Codec.Serialise.deserialiseOrFail bytesLazy)
+            normalisedImport <- Dhall.Core.throws (Dhall.Binary.decodeExpression term)
+            -- a frozen import does not have any dependencies if loaded from the semantic cache.
+            let graph = Tree.Node import_ []
+            -- TODO: cached imports should have type Expr Src X!
+            return (SemanticImport {normalisedImport = fmap undefined normalisedImport, ..})
+
+        Nothing -> do  -- continue with semi-semantic cache
+            SemanticImport { normalisedImport = normalisedImport, graph = Tree.Node _ imports }
+                <- processImportWithSemisemanticCache stack import_
+
+            -- The semi-semantic cache always returns a semantic hash computed
+            -- with the default `StandardVersion`. In general we will need to
+            -- check all possible versions to find the one that matches the
+            -- requested semantic hash, if it exists.
+            let variants = map (\version -> encodeExpression version normalisedImport)
+                               [ minBound .. maxBound ]
+            case Data.Foldable.find ((== semanticHash). Crypto.Hash.hash) variants of
+                Just bytes -> lift $ writeToSemanticCache semanticHash bytes
+                Nothing -> do
+                    let expectedHash = semanticHash
+                    version <- use standardVersion
+                    let actualHash = hashExpression version normalisedImport
+                    throwMissingImport (Imported stack (HashMismatch {..}))
+
+            return (SemanticImport { graph = Tree.Node import_ imports, .. })
+
+-- Try to load from "semi-semantic" disk cache, otherwise typecheck and
+-- normalise from scratch.
```

Comment:
You might want to take your description of the semi-semantic cache from https://github.com/dhall-lang/dhall-haskell/issues/1098 and include it in the documentation for this function

Reply:
Good point.